src/OpenTelemetry.Extensions.Enrichment/BaseEnrichmentBag.cs

```diff
@@ -0,0 +1,36 @@
+// <copyright file="BaseEnrichmentBag.cs" company="OpenTelemetry Authors">
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// </copyright>
+
+using System.Collections.Generic;
+
+namespace OpenTelemetry.Extensions.Enrichment;
+
+public abstract class BaseEnrichmentBag
```

Comment:
We anyways have to have 3 different types for Trace/Log/Metric enrichment bags so could we use structs here so we can remove the pooling and ends up with better perf. What about something like this

```csharp
public struct TraceEnrichmentBag
{
    public TraceEnrichmentBag(Activity activity)
    {
        Activity = activity;
    }
    public Activity Activity { get; init; }
    public void Add(string key, object? value)
    {
        Activity.AddTag(key, value);
    }
}

public struct MetricEnrichmentBag
{
    public MetricEnrichmentBag(Instrument instrument)
    {
        TagList = new TagList();
        Instrument = instrument;
    }
    internal TagList TagList { get; init; } // This is a internal placeholder and we could implement our own struct here later
    public Instrument Instrument { get; init; }
    public void Add(string key, object? value)
    {
        TagList.Add(key, value);
    }
}

public abstract class TraceEnricher
{
    public virtual void Enrich(ref TraceEnrichmentBag traceEnrichmentBag)  {  }
}
```
Then the need for pools are eliminated completely, will be better perf wise
```csharp
public class TraceEnrichmentProcessor : BaseProcessor<Activity>
{
    public void OnEnd(Activity activity)
    {
        var enrichmentBag = new TraceEnrichmentBag(activity);
        foreach(var enricher in _enrichers) { enricher.Enrich(enrichmentBag); }
    }
}
```

Reply:
@CodeBlanch What do you think?

Reply:
I updated the code